Add the RISC-V entry point assembly to libtock_runtime

jrvanwhy opened the pull request:
To support the entry point assembly, I rewrote the linker script ([link to the new version](https://github.com/jrvanwhy/libtock-rs/blob/riscv-asm/core/runtime/layout_generic.ld) in case you don't want to see the diff).

The rust toolchain is unable to compile raw assembly files on its own. I don't want a separate toolchain to be a build requirement of `libtock-rs`, so instead I committed the compiled library into the repository. To keep the compiled library and source code in sync, I added an actions workflow that verifies the compiled library matches the provided sources.

There is one part of this PR I am unhappy with. I have to manually `strip` local symbols from the compiled library -- i.e., removing the `riscv64-linux-gnu-strip` invocation from `assemble.sh` gives the following:

```
Disassembly of section .start:

00000000 <start>:
   0:   00000417                auipc   s0,0x0
   4:   87aa                    mv      a5,a0
   6:   4384                    lw      s1,0(a5)
   8:   00940c63                beq     s0,s1,20 <.Lset_brk>
   c:   4521                    li      a0,8
   e:   4585                    li      a1,1
  10:   4609                    li      a2,2
  12:   4709                    li      a4,2
  14:   00000073                ecall
  18:   4501                    li      a0,0
  1a:   4719                    li      a4,6
  1c:   00000073                ecall

00000020 <.Lset_brk>:
  20:   4501                    li      a0,0
  22:   43cc                    lw      a1,4(a5)
  24:   4715                    li      a4,5
  26:   00000073                ecall
  2a:   0087a103                lw      sp,8(a5)
  2e:   47c8                    lw      a0,12(a5)
  30:   c909                    beqz    a0,42 <.Lzero_bss>
  32:   4b8c                    lw      a1,16(a5)
  34:   4bd0                    lw      a2,20(a5)

00000036 <.Ldata_loop_body>:
  36:   4194                    lw      a3,0(a1)
  38:   c214                    sw      a3,0(a2)
  3a:   1571                    addi    a0,a0,-4
  3c:   0591                    addi    a1,a1,4
  3e:   0611                    addi    a2,a2,4
  40:   f97d                    bnez    a0,36 <.Ldata_loop_body>

00000042 <.Lzero_bss>:
  42:   4f88                    lw      a0,24(a5)
  44:   c519                    beqz    a0,52 <.Lcall_rust_start>
  46:   4fcc                    lw      a1,28(a5)

00000048 <.Lbss_loop_body>:
  48:   00058023                sb      zero,0(a1)
  4c:   157d                    addi    a0,a0,-1
  4e:   0585                    addi    a1,a1,1
  50:   fd65                    bnez    a0,48 <.Lbss_loop_body>

00000052 <.Lcall_rust_start>:
  52:   fafff0ef                jal     ra,0 <start>
```

I'd appreciate advice on how to prevent `riscv64-linux-gnu-as` from including the local symbols in its output.

mcy reviewed (commented):
Left some comments in the asm; the linker script LGTM.

Question: is there a reason that `global_asm!` is an unacceptable solution? I'm... slightly grossed out by the idea of checking in a binary so I want to see if there isn't a workaround.

jrvanwhy commented:
> Question: is there a reason that `global_asm!` is an unacceptable solution? I'm... slightly grossed out by the idea of checking in a binary so I want to see if there isn't a workaround.

AFAIK, there are four workable solutions for the entry point assembly (which I'll order from "least gross" to "grossest", as judged by me):

A. The `naked_functions` unstable feature
B. The `global_asm` unstable feature
C. Build the binary in `build.rs`
D. Check in a binary

The last time the Tock core WG discussed moving to stable Rust, the consensus was that Tock should try to move to stable Rust when `asm` is stabilized, but the group wasn't 100% sure that doing so was possible. I suggested that we use `libtock-rs` to evaluate the feasibility of that plan. As such, I am trying to avoid the use of any unstable features except `asm`. In particular, `asm` will probably be stabilized this year (or soon thereafter), while neither `global_asm` nor `naked_functions` appear to be nearing stabilization.

I chose to check in a binary rather than build the binary in `build.rs` because building the library in `build.rs` requires users to have a RISC-V toolchain installed, and they can be a lot of work to set up. We can't really write setup instructions because at this point the instructions would have to vary per distribution. E.g. we're using a 64-bit RISC-V toolchain in GitHub Actions because that's available.

I believe the GitHub Actions workflow I included in this PR minimizes the drawbacks of checking in a binary. Actually, writing this reminded me I need to add the new workflow to `bors.toml` so its failure blocks PR merges.

jrvanwhy commented on core/runtime/asm/asm_riscv32.S:
```diff
@@ -0,0 +1,57 @@
+	li a0, 0  /* exit-terminate */
+	/* TODO: Set a completion code, once completion codes are decided */
+	li a4, 6  /* `exit` class */
+	ecall
```
I requested that `exit-terminate` never return specifically to avoid needing an infinite loop here. I would prefer to avoid adding to the binary size to address a situation that can only occur with a broken kernel.

mcy replied:
Alright, sounds good.

jrvanwhy commented on core/runtime/asm/asm_riscv32.S:
```diff
@@ -0,0 +1,57 @@
+	lw sp, 8(a5)  /* sp = rt_header._stack_top */
+
+	/* Copy .data into place. */
+	lw a0, 12(a5)              /* remaining = rt_header.data_size */
```
I'm having trouble getting that working. I added `#define START_OFFSET 0` to the top of the file, and changed the `lw s1, 0(a5)` line to `lw s1, START_OFFSET(a5)`. I got the following error:

```
asm_riscv32.S: Assembler messages:
asm_riscv32.S:11: Error: illegal operands `lw s1,START_OFFSET(a5)'
```

I'm having trouble finding documentation on how exactly `#define` works in assembly. Any idea what I'm doing wrong?

mcy replied:
I was pretty sure it behaves exactly like the prepossessor?  If you can't get it to work feel free to ignore.

jrvanwhy replied:
My initial thought is that it looks like call synax, so maybe it's passing in `a5` as a parameter -- but [rust-embedded/riscv-rt](https://github.com/rust-embedded/riscv-rt/blob/47ece5f5163a2e38ce5e4685b5d3145713d7954a/asm.S#L134) uses the same syntax and evidently it works.

I can't get it to work :-(

jrvanwhy commented on core/runtime/asm/asm_riscv32.S:
```diff
@@ -0,0 +1,57 @@
+	li a4, 6  /* `exit` class */
+	ecall
+
+	.Lset_brk:
```
Done. I have preference because I'm still an assembly newbie with no style.

jrvanwhy commented on core/runtime/asm/asm_riscv32.S:
```diff
@@ -0,0 +1,57 @@
+	.Ldata_loop_body:
+	lw a3, 0(a1)               /* a3 = *src */
+	sw a3, 0(a2)               /* *dest = a3 */
+	c.addi a0, -4              /* remaining -= 4 */
```
I have no idea why I did that. Changed to a normal `addi`.

jrvanwhy commented on core/runtime/asm/asm_riscv32.S:
```diff
@@ -0,0 +1,57 @@
+	bnez a0, .Lbss_loop_body    /* Iterate again if remaining != 0 */
+
+	.Lcall_rust_start:
+	jal rust_start
```
`rust_start` is not in this PR, but it's prototype is `extern fn rust_start() -> !`, so it really should not be able to return. I added a comment indicating that it should never return. Is that good enough?

mcy replied:
SG!

jrvanwhy commented on core/runtime/layout_generic.ld:
```diff
@@ -1,171 +1,124 @@
+    /* Runtime header. Contains values the linker knows that the runtime needs
+     * to look up.
+     */
+    .rt_header : {
```
Not really, I consider this the definition. I do have a separate piece of code (that I do not want to add to this PR because this PR is large enough already). That code isn't intended to be part of `libtock-rs`, though, just a tool to help develop it (it is a Rust version of `start` you can copy + paste assembly from).

alistair23 commented:
We shouldn't be checking in a binary. When a user edits some code and builds that code should be built. With this change we are ignoring their code and just linking with an existing binary. That is a bad idea.

Building the assembly in `build.rs` is an ok option, but I would say using either `naked_functions` or `global_asm` is a better option. It's not stable, but I don't see why that is such a problem. The kernel is built with unstable Rust so we aren't diverging from that.

jrvanwhy commented:
> Building the assembly in `build.rs` is an ok option, but I would say using either `naked_functions` or `global_asm` is a better option. It's not stable, but I don't see why that is such a problem. The kernel is built with unstable Rust so we aren't diverging from that.

I *strongly* prefer to avoid `naked_functions` and `global_asm`.

I am okay with building the assembly in `build.rs`. The main drawback is it would require all `libtock-rs` contributors to have both `binutils-arm-none-eabi` and `binutils-riscv64-linux-gnu` installed (my dev environment is Debian 10, so I don't have a `riscv32` toolchain, so this is all I can support). If both you and @hudson-ayers are okay with that constraint then I will make that change.

alistair23 commented:
What is the problem with `naked_functions` or `global_asm`?

It doesn't seem unreasonable that someone doing embedded development should install binutils. It does somewhat go against the ease of using Rust though, which is a downside.

jrvanwhy commented:
> What is the problem with `naked_functions` or `global_asm`?

Because I want at least `libtock-rs` (and hopefully also the Tock kernel) to move to stable Rust in the next year. Neither of those are on track for stabilization in that timeframe.

hudson-ayers commented:
`libtock-c` already requires `gcc-arm-none-eabi`, which depends on `binutils-arm-none-eabi`, and *optionally* requires `gcc-riscv64-unknown-elf`, which depends on `binutils-riscv64-unknown-elf`.

It would probably be nice to require the same riscv toolchain for both userspace libraries, as I think there are plenty of people that use both. It would also let us re-use the installation instructions. Functionally I am not sure the difference between `riscv64-unknown-elf` and `riscv64-linux-gnu`

The way `libtock-c` makes the risc-v toolchain optional is by requiring an environment variable be set to turn on risc-v builds by default. I don't personally feel strongly either way about whether we should do something similar in `libtock-rs`. Overall I think the `build.rs` approach is fine

jrvanwhy commented:
Debian 10 has `riscv64-linux-gnu` but not `riscv64-unknown-elf`, while Ubuntu 20.04 has both. I think we have the following possibilities:

A. Depend on `riscv64-linux-gnu`, which is distinct from `libtock-c`.
B. Have `build.rs` try `riscv64-unknown-elf` and fallback to `riscv64-linux-gnu` if it is not available. Use `riscv64-unknown-elf` in GitHub Actions.

I don't have a preference between the two options.

alistair23 commented:
I still prefer `naked_functions`, it doesn't look like they are too far behind `inline_asm` https://github.com/rust-lang/rust/issues/32408#issuecomment-484726301

If using `build.rs` then option B is the better option. The Linux toolchain should only be a last resort for users that don't install the baremetal toolchain.

hudson-ayers commented:
Unfortunately after looking some more into `naked_fn` stabilization timeline, I found this more recent comment by the same individual, on the rust zulip (https://rust-lang.zulipchat.com/#narrow/stream/216763-project-inline-asm/topic/Next.20steps.20for.20the.20group/near/197562883):

>Amanieu 7:27 AM, May 14 2020:
> I'm particularly interested in bringing global_asm! up to speed with the new features of asm!. As a bonus, this would allow us to deprecate the existing naked function support and replace it with a proc macro.

I believe we have a desire to not use procedural macros in `libtock-rs`, as they add several dependencies, so I am not sure we can be confident `naked_fn` will be a long-term solution.

hudson-ayers commented:
Option B sounds good to me, but we should be careful that whatever we end up with works for developers on mac as well (assuming they follow the riscv toolchain instructions currently in `libtock-c`).

jrvanwhy commented:
I've changed it to build the external assembly from libtock-rs, and removed the binary. This had several follow-on effects:

1. I discovered that the linker script selection is *still* broken. Evidently linker script includes are global rather than relative, and the `INCLUDE layout_generic.ld` lines in the per-board linker scripts was including `libtock-rs`' `layout_generic.ld` rather than `libtock_runtime`'s `layout_generic.ld`. To avoid name collisions with application code, I renamed the generic layout to `libtock_layout.ld`.
1. `build.rs` now correctly fails when you try to build `libtock_runtime` on unsupported architectures (e.g. `x86_64`). The `clippy` and `miri` invocations that `make test` does started failing. I reworked these invocations to check as many things as possible without failing.
1. I updated the actions environments to Ubuntu 20.04 because Ubuntu 18.04 does not have the toolchain we need.
1. Last, I added a Mac OS workflow to see if it works on Mac OS. I expect this to take many tries to get right, so bear with me while I debug it live (I have never developed on Mac OS before, and don't have a Mac around to test on).

alistair23 commented on core/runtime/asm/asm_riscv32.S:
```diff
@@ -0,0 +1,58 @@
+.section .start, "ax"
+.globl start
+start:
```
Could we add a comment describing the arguments?

alistair23 replied:
Also probably worth commenting what `rt_header` offsets are, just to make things clear

jrvanwhy replied:
I added a top-of-file comment describing `rt_header` and a "function comment" describing the arguments.

alistair23 commented on core/runtime/extern_asm.rs:
```diff
@@ -0,0 +1,133 @@
+        "riscv32" => &[
+            // First try riscv64-unknown-elf, as it is the toolchain used by
+            // libtock-c.
+            AsmBuildConfig {
```
Can we add a `riscv32-unknown-elf` without the `as_extra_args` as well

jrvanwhy replied:
Do you have a `riscv32-unknown-elf` toolchain? If yes then I'll probably put it first, as it seems to be the "correct" toolchain for the task. If not I'll put it last because it will be the only untested toolchain.

jrvanwhy replied:
Actually, if you have `riscv32-unknown-elf` maybe I should put it in the middle, because it still won't be tested by our CI.

alistair23 replied:
I have a `riscv32-oe-elf` toolchain, which is a vendored version of `riscv32-unknown-elf`. I don't think we should add that, but I suspect some distros will have a 32-bit only toolchain.

I would say put it in the middle.

jrvanwhy replied:
I added `riscv32-unknown-elf` between `riscv64-unknown-elf` and `riscv64-linux-gnu`.

hudson-ayers commented on core/runtime/libtock_layout.ld:
```diff
@@ -0,0 +1,124 @@
+     */
+    .stack (NOLOAD) : {
+        KEEP(*(.stack_buffer))
+        _stack_top = ALIGN(16);  /* Used in rt_header */
```
in the old linker file, we both aligned the stack using `ALIGN`, and checked the size was an 8 byte multiple at the end of the file using an `ASSERT`:
```
ASSERT((_stack_top_aligned - _stack_top_unaligned) == 0, "
STACK_SIZE must be 8 byte multiple")
```

I seem to recall from the conversation [here](https://github.com/tock/libtock-rs/pull/245#discussion_r505726950) that the assert was necessary for friendly debug messages in case someone writes an app with an inappropriate stack size. Has that changed, or should we keep that assert around so that someone who passes an invalid stack size can understand their error?

jrvanwhy replied:
Actually, what I had here was broken. It could potentially overlap `.data` with the top of the stack.

We're going to need alignment padding at some point. We can either align `.stack` and then `ASSERT` that the stack size must be a multiple of 16 bytes, or just align the top of stack (and leave the `.stack` section unaligned). I chose to do the latter, which I did in the latest commit. Therefore there is no more need for the `ASSERT`.

hudson-ayers replied:
Makes sense, thanks

hudson-ayers reviewed (commented):
This looks good, thanks for taking the time to set up testing on Mac and support the `libtock-c` version of the toolchain.

A few comments:
- good catch on MPU_MIN_ALIGN being unused
- is there a reason TBF_HEADER_SIZE is specified per board? I guess that once we are adding permissions to the header the size might vary depending on the board?
- good catch on 16 byte stack alignment for risc-v, had we just gotten lucky that our previous 8-byte alignment didn't cause issues?

jrvanwhy commented:
> * is there a reason TBF_HEADER_SIZE is specified per board? I guess that once we are adding permissions to the header the size might vary depending on the board?

I didn't think about this. This should probably be a per-process-binary value like stack size instead of a per-board value.

It's a bit difficult to test right now, as this PR doesn't contain a process binary to test with. I added a TODO to look at making this per-process-binary in the future (and yes, I do intend to go back through and handle my TODOs later, most of them are waiting on code that hasn't been merged upstream yet).

> * good catch on 16 byte stack alignment for risc-v, had we just gotten lucky that our previous 8-byte alignment didn't cause issues?

I believe so.

jrvanwhy commented on .github/workflows/mac_os.yml:
```diff
@@ -0,0 +1,39 @@
+
+      # Install the toolchains we need, the run the Makefile's test action. We
+      # let the makefile do most of the work because the makefile can be tested
+      # locally. Using -j2 because the Actions VMs have 2 cores.
```
Fixed.

jrvanwhy commented on .github/workflows/mac_os.yml:
```diff
@@ -0,0 +1,39 @@
+      - name: Clone repository
+        uses: actions/checkout@v2.3.0
+        with:
+          submodules: true
```
Cleaned up, thanks.

jrvanwhy commented:
I re-ran the Mac workflow but with each command broken out into a separate step so GitHub gives me the times ([run results](https://github.com/jrvanwhy/libtock-rs/pull/16/checks?check_run_id=1842629610)). It took 72 minutes in total, of which 68 were spent on the `brew install riscv-gnu-toolchain --with-multilib` step.

If anyone who uses Mac OS knows a faster way to install `binutils-riscv64-unknown-elf`, that would be appreciated!

hudson-ayers commented:
I know brad opened [this issue](https://github.com/riscv/homebrew-riscv/issues/22) to no avail. I am not sure if there can be some improvement from the fact you only need that single tool. 

Alternatively, maybe this would be a good place to use GA caching, but I don't think its fair to require that of you and would probably recommend just removing the macOS CI on PRs and instead just run it on merges into master (so we would get after-the-fact notifications about breakages without slowing most people's development flow)

jrvanwhy commented:
> Alternatively, maybe this would be a good place to use GA caching, but I don't think its fair to require that of you

On the surface, GA caching sounds like it a solution -- but it's unclear to me what directories/files we would need to cache. If they're scattered everywhere then it wouldn't be very practical. I agree that I am not the person to do it, as my only access to Mac OS is through GitHub Actions...

> and would probably recommend just removing the macOS CI on PRs and instead just run it on merges into master (so we would get after-the-fact notifications about breakages without slowing most people's development flow)

We could let it run on PRs but stop having Bors run it, so it doesn't slow down PR merges. That way we would typically catch issues before merging to master. May need to add a note to our code review document telling people not to wait for the Mac build.

hudson-ayers commented:
+1 for not having bors run it

jrvanwhy commented:
> +1 for not having bors run it

I made that change. I also renamed it to `mac-os` from `mac_os` to be consistent with `size-diff`.

bradjc commented:
I tried asking about having the 32 bit tools pre-compiled, but that went nowhere: https://github.com/riscv/homebrew-riscv/issues/22#issuecomment-643840201

The with --multilib adds support for 32 bit, and with brew requires a full compile of the toolchain.

jrvanwhy commented:
> I tried asking about having the 32 bit tools pre-compiled, but that went nowhere: [riscv/homebrew-riscv#22 (comment)](https://github.com/riscv/homebrew-riscv/issues/22#issuecomment-643840201)
> 
> The with --multilib adds support for 32 bit, and with brew requires a full compile of the toolchain.

Thank you for confirming!

jrvanwhy commented:
This PR is ready to merge now.

hudson-ayers commented:
bors r+
